Add mypy support

jaraco opened the pull request:
https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3399

jaraco commented:
[This test run](https://app.circleci.com/pipelines/github/tahoe-lafs/tahoe-lafs/1496/workflows/77d849e7-e809-488b-bc12-718e99ee3ca2/jobs/38640) demonstrates the checks now passing in CI.

exarkun commented on src/allmydata/uri.py:
```diff
@@ -701,7 +712,7 @@ class DirectoryURIVerifier(_DirectoryBaseURI):
     BASE_STRING=b'URI:DIR2-Verifier:'
     BASE_STRING_RE=re.compile(b'^'+BASE_STRING)
-    INNER_URI_CLASS=SSKVerifierURI
+    INNER_URI_CLASS=SSKVerifierURI  # type: Type[IVerifierURI]
```
I'm not sure I know what this means.  Are there some mypy or mypy-zope docs I should read?  Neither the docstring for `typing.Type` nor the information on https://realpython.com/python-type-checking/ seems to explain what's happening here.

jaraco replied:
I don't have any good references. I've just been picking it up by trial-and-error. What I think this is saying is the `INNER_URI_CLASS` is a `Type` (aka class) that's an instance of the metaclass `IVerifierURI`. That's only really a guess, though. I admit, this approach may not be correct, but it's one solution that unblocks the errors that occur when subclasses of DirectoryURIVerifier indicate a different type for INNER_URI_CLASS. The intention here is to declare some type that's common among all DirectoryURIVerifiers.

exarkun commented on src/allmydata/web/private.py:
```diff
@@ -61,7 +61,11 @@ def check(auth_token):
 
 
-@implementer(IToken)
+# Shoobx/mypy-zope#26
```
Shoobx/mypy-zope#26 is closed now.  It sounds like there is some other resolution required here, perhaps having to do with using some type stubs for Twisted?  Alternatively, if you want to move `IToken` to `allmydata.interfaces` and that helps then feel free.

Either way, please update the comment to explain the pitfall and link to any outstanding tickets for work that might need to be done to address it.

jaraco replied:
That's right - having type stubs for Twisted is the proper fix. Clarified in 51b0b201b.

exarkun commented on src/allmydata/windows/fixups.py:
```diff
@@ -217,7 +217,12 @@ def writelines(self, lines):
-        return re.sub(u'\\x7F[0-9a-fA-F]*\\;', lambda m: unichr(int(m.group(0)[1:-1], 16)), s)
+        return re.sub(
+            u'\\x7F[0-9a-fA-F]*\\;',
+            # type ignored for 'unichr'
```
maybe mention it is ignored because it doesn't exist on Python 3?  I assume this (and other Python 2/3 ignores) will go away as the porting effort reaches this code.  It might be nice to help the porter who gets to such code remember to remove the ignore somehow.

jaraco replied:
Fixed in efd0aef28.

exarkun reviewed (changes requested):
Thanks.  Left a bunch of comments inline.  Some additional thoughts:

* There are some conflicts with master now, alas.
* CI isn't completely happy.
* Are we going to need a ratchet-style job for mypy checks at some point?  It seems like at the end of *this* PR we have mypy running with no errors but, eg, are a ton of errors going to appear when we bring in the type stubs for our dependencies?  Since that's later it doesn't need to be a blocker for this PR, I'm just wondering - and if there's anything we can do now to make our lives easier then it might be worth it.

jaraco commented on .circleci/config.yml:
```diff
@@ -475,6 +478,17 @@ jobs:
+  typechecks:
+    docker:
+      - <<: *DOCKERHUB_AUTH
+        image: "jaraco/multipy-tox"
```
Agreed. There's nothing special about this image except that it has tox on a late Python 3. Probably the existing images can work - they might need some tweaks to support this. I was hoping someone familiar with those images could figure out the recipe after I had something stable on jaraco/multipy-tox, so yes, I'd appreciate the help.

exarkun replied:
Pushed a change that switches this to one of our Ubuntu images that has Python 3 and tox.  Here's an example CI run:

https://app.circleci.com/pipelines/github/LeastAuthority/tahoe-lafs/154/workflows/2dd439d9-514a-4cd3-bc8d-2ccbe6ea8cb9/jobs/2485

jaraco replied:
Change is a01078ddec86e1d193c7940087d6506d376028d8. Looks good! Thanks for doing that.

jaraco commented on tox.ini:
```diff
@@ -108,6 +108,16 @@ commands =
+deps =
+    mypy
+    git+https://github.com/jaraco/mypy-zope@bugfix/21-InterfaceClass-subclass
+    git+https://github.com/jaraco/foolscap@bugfix/75-use-metaclass
```
I definitely share that preference. I saw it as a non-blocker, as the checks can have their benefit without the patches, but I would definitely like to see these merged, and making it a blocker will incentivize getting those merged and released.

jaraco replied:
Pull requests:
- warner/foolscap#74
- warner/foolscap#76
- Shoobx/mypy-zope#24

jaraco replied:
Now that those are merged and c7a4cdb44d801034de1e068f10aa3d8df8a4950f is committed, the project now relies on the main branch of the canonical repo.

jaraco commented on src/allmydata/__init__.py:
```diff
@@ -14,7 +14,7 @@
 __version__ = "unknown"
 try:
-    from allmydata._version import __version__
+    from allmydata._version import __version__  # type: ignore
```
Hmm. Perhaps mypy improved these cases. I searched and didn't find anything about it. I tried and can't find a way to replicate whatever was happening, so I'll revert this change.

jaraco replied:
Removed in 950ca1893

jaraco replied:
Turns out the [tests fail in CI](https://app.circleci.com/pipelines/github/tahoe-lafs/tahoe-lafs/1647/workflows/60ae95d4-abe8-492c-8a03-1ad3b9e42ed3/jobs/40972) but not locally. Not sure why the difference. Regardless, I'll add a comment.

jaraco replied:
Brought back in b65ef3cee with a comment showing where it fails.

jaraco commented on src/allmydata/frontends/ftpd.py:
```diff
@@ -317,7 +315,7 @@ def logout(): pass
 class FTPServer(service.MultiService):
     def __init__(self, client, accountfile, accounturl, ftp_portstr):
-        precondition(isinstance(accountfile, (unicode, NoneType)), accountfile)
+        precondition(isinstance(accountfile, (unicode, type(None))), accountfile)
```
I definitely aim to minimize the impact on behavior, so I'll stick with this change for now, but I agree it may be preferable to use the idiomatic form separately.

jaraco commented on src/allmydata/test/test_dirnode.py:
```diff
@@ -1526,7 +1526,7 @@ def test_deep_immutable(self):
 
 @implementer(IMutableFileNode)
-class FakeMutableFile(object):
+class FakeMutableFile(object):  # type: ignore # incomplete interface
```
Right. I think what I was trying to say (succinctly enough to fit within the line length limit) was that the interface of this class is incomplete with respect to its declared interface. Fixed in 99da74fff.

jaraco commented on src/allmydata/uri.py:
```diff
@@ -535,6 +540,12 @@ def get_verify_cap(self):
+
+    def is_readonly(self):
+        raise NotImplementedError()
+
```
Indeed. I don't know either. In other cases, I went the other route. It's a little annoying that every subclass would need to have redundant declarations (interface + subclass), but that's probably a better approach than implementing/not-implementing the interface in the parent class (as seen here).

jaraco replied:
Fixed in 6b6b8f837. PTAL.

jaraco replied:
That commit was insufficient. Amended with ab2c544efcaad629f2d49173bfdc82e8de94d944.

jaraco commented on src/allmydata/codec.py:
```diff
@@ -57,6 +57,10 @@ def encode(self, inshares, desired_share_ids=None):
         return defer.succeed((shares, desired_share_ids))
 
+    def encode_proposal(self, data, desired_share_ids=None):
+        raise NotImplementedError()
```
It looks like `encode_proposal` is a proposed but not-yet-implemented replacement for `encode`. See the docstrings in interfaces.py, especially these:
https://github.com/tahoe-lafs/tahoe-lafs/blob/d19b1cfd687cfede9374d994aca7c1890af6811e/src/allmydata/interfaces.py#L1706-L1716
https://github.com/tahoe-lafs/tahoe-lafs/blob/d19b1cfd687cfede9374d994aca7c1890af6811e/src/allmydata/interfaces.py#L1724-L1726

Renaming the method name in the interface isn't possible as the interface definition for encode also exists and provides a different interface.

I believe this implementation as proposed is the cleanest given the proposal has apparently not been accepted. I would recommend removing this proposed interface, replacing it with an issue or PR to capture the work in progress, but independent of this PR.

jaraco commented on src/allmydata/immutable/encode.py:
```diff
@@ -711,3 +711,6 @@ def get_uri_extension_data(self):
         return self.uri_extension_hash
+
+    def set_size(self, size):
+        raise NotImplementedError()
```
Fixed in ea0c10ef8.

jaraco commented on src/allmydata/immutable/offloaded.py:
```diff
@@ -499,10 +499,12 @@ def close(self):
         return self.call("close")
 
+    def set_upload_status(self, upload_status):
+        raise NotImplementedError
```
Fixed in d051791e9.

jaraco commented on src/allmydata/immutable/upload.py:
```diff
@@ -385,6 +385,12 @@ def get_share_placements(self):
+
+    def confirm_share_allocation(self, peerid, shnum):
+        raise NotImplementedError
+
```
Fixed in 090031cbf.

jaraco commented on src/allmydata/introducer/client.py:
```diff
@@ -198,7 +198,7 @@ def create_announcement_dict(self, service_name, ann):
         return ann_d
 
-    def publish(self, service_name, ann, signing_key):
+    def publish(self, service_name, ann, signing_key=None):
```
Fixed in 0e248cb4e.

jaraco commented on src/allmydata/mutable/filenode.py:
```diff
@@ -951,7 +951,7 @@ def get_size(self):
 
 
-    def download_to_data(self, fetch_privkey=False, progress=None):
+    def download_to_data(self, fetch_privkey=False, progress=None):  # type: ignore # fixme
```
Here's what I see:

```
tahoe-lafs 3399.mypy $ git diff
diff --git a/src/allmydata/mutable/filenode.py b/src/allmydata/mutable/filenode.py
index 4613a918b..e9cf23fa1 100644
--- a/src/allmydata/mutable/filenode.py
+++ b/src/allmydata/mutable/filenode.py
@@ -951,7 +951,7 @@ class MutableFileVersion(object):
         return self._servermap.size_of_version(self._version)
 
 
-    def download_to_data(self, fetch_privkey=False, progress=None):  # type: ignore # fixme
+    def download_to_data(self, fetch_privkey=False, progress=None):
         """
         I return a Deferred that fires with the contents of this
         readable object as a byte string.

tahoe-lafs 3399.mypy $ tox -e typechecks
typechecks installed: attrs==20.3.0,Automat==20.2.0,cffi==1.14.4,constantly==15.1.0,cryptography==3.3.1,foolscap @ git+https://github.com/jaraco/foolscap@d70d8ecee926d3693b9fac55aebe13573d4f71c9,hyperlink==20.0.1,idna==2.10,incremental==17.5.0,mypy==0.790,mypy-extensions==0.4.3,mypy-zope @ git+https://github.com/jaraco/mypy-zope@48f912cc3cdb4ee37d4d3783ef466a7dce8bf0a9,pyasn1==0.4.8,pyasn1-modules==0.2.8,pycparser==2.20,PyHamcrest==2.0.2,pyOpenSSL==20.0.1,service-identity==18.1.0,six==1.15.0,Twisted==20.3.0,typed-ast==1.4.1,typing-extensions==3.7.4.3,zope.event==4.5.0,zope.interface==5.2.0,zope.schema==6.0.0
typechecks run-test-pre: PYTHONHASHSEED='1134400167'
typechecks run-test: commands[0] | mypy src
src/allmydata/mutable/filenode.py:954: error: Signature of "MutableFileVersion" incompatible with "download_to_data" of supertype "allmydata.interfaces.IReadable"
Found 1 error in 1 file (checked 293 source files)
ERROR: InvocationError for command /Users/jaraco/code/public/tahoe-lafs/.tox/typechecks/bin/mypy src (exited with code 1)
________________________________________________________________ summary ________________________________________________________________
ERROR:   typechecks: commands failed
```

jaraco commented on src/allmydata/mutable/filenode.py:
```diff
@@ -1205,3 +1205,6 @@ def _update_servermap(self, mode=MODE_WRITE, update_range=None):
         return u.update()
+
+    def get_servermap(self):
+        raise NotImplementedError
```
Fixed in c2d2aba83.

jaraco commented on src/allmydata/node.py:
```diff
@@ -713,7 +713,7 @@ class Node(service.MultiService):
     NODETYPE = "unknown NODETYPE"
     CERTFILE = "node.pem"
-    GENERATED_FILES = []
+    GENERATED_FILES = []  # type: ignore
```
Sure. I'd observed the same, but gave the code the benefit of the doubt and left it around in case it was used through some mysterious usage I couldn't detect. Fixed in 189608e11.

jaraco commented on src/allmydata/scripts/common.py:
```diff
@@ -64,7 +69,7 @@ class BasedirOptions(BaseOptions):
         ["basedir", "C", None, "Specify which Tahoe base directory should be used. [default: %s]"
          % quote_local_unicode_path(_default_nodedir)],
-    ]
+    ]  # type: List[Sequence[Any]]
```
Great idea. Fixed in 602a06e5c.

jaraco commented on src/allmydata/scripts/common.py:
```diff
@@ -101,7 +106,7 @@ class NoDefaultBasedirOptions(BasedirOptions):
     optParameters = [
         ["basedir", "C", None, "Specify which Tahoe base directory should be used."],
-    ]
+    ]  # type: List[Sequence[Any]]
```
602a06e5c

jaraco commented on src/allmydata/scripts/runner.py:
```diff
@@ -108,7 +113,7 @@ def postOptions(self):
 create_dispatch = {}
 for module in (create_node, stats_gatherer):
-    create_dispatch.update(module.dispatch)
+    create_dispatch.update(module.dispatch)  # type: ignore
```
I wondered the same thing. I suspect it's because `(create_node, stats_gatherer)` resolves to `Tuple[ModuleType]` where `ModuleType` is the common interface of the `create_node` and `stats_gatherer` modules and `ModuleType` doesn't generally have a `dispatch` attribute. My guess is one would need to create an interface. to satisfy this check, and that's not simple because it would also require creating a type for the dispatch property as well.

jaraco commented on src/allmydata/stats.py:
```diff
@@ -78,7 +78,7 @@ def get_stats(self):
 class CPUUsageMonitor(service.MultiService):
     HISTORY_LENGTH = 15
-    POLL_INTERVAL = 60
+    POLL_INTERVAL = 60  # type: float
```
This was surprising to me too. I learned that `int` is a subtype of `float`, so if you want to declare a number that can be an integer or float, you just declare it as a float. I would recommend against adding `.0` or `.00`. Integers are acceptable here and should be used where convenient and appropriate.

jaraco commented:
Thanks for the insightful review. I've addressed some of the concerns but have many more to go. I'm taking a break and will resume later.

jaraco commented on src/allmydata/test/storage_plugin.py:
```diff
@@ -48,7 +48,7 @@ def just_some_method():
 
 
-@implementer(IFoolscapStoragePlugin)
+@implementer(IFoolscapStoragePlugin)  # type: ignore # todo: make stubs for twisted
```
In dacdf7f12, I added a link to the existing upstream issue.

jaraco commented on src/allmydata/storage/server.py:
```diff
@@ -581,7 +581,7 @@ def _add_or_renew_leases(self, shares, lease_info):
             share.add_or_renew_lease(lease_info)
 
-    def slot_testv_and_readv_and_writev(
+    def slot_testv_and_readv_and_writev(  # type: ignore # fixme
```
I agree, it probably does have to do with that issue. I think I'd missed that detail. Fixed in 5396f9f97.

jaraco commented on src/allmydata/test/no_network.py:
```diff
@@ -67,7 +67,7 @@ class Marker(object):
 fireNow = partial(defer.succeed, None)
 
-@implementer(IRemoteReference)
+@implementer(IRemoteReference)  # type: ignore # todo: write stubs for foolscap
```
Done in 1bf71fd69.

jaraco commented on src/allmydata/test/no_network.py:
```diff
@@ -258,7 +261,7 @@ def create_no_network_client(basedir):
 
 
-class _NoNetworkClient(_Client):
+class _NoNetworkClient(_Client):  # type: ignore # Cannot determine consistent MRO order
```
Here's what I see:

```
tahoe-lafs 3399.mypy $ git diff
diff --git a/src/allmydata/test/no_network.py b/src/allmydata/test/no_network.py
index ba497f81a..17a63a3a2 100644
--- a/src/allmydata/test/no_network.py
+++ b/src/allmydata/test/no_network.py
@@ -261,7 +261,7 @@ def create_no_network_client(basedir):
     return defer.succeed(client)
 
 
-class _NoNetworkClient(_Client):  # type: ignore # Cannot determine consistent MRO order
+class _NoNetworkClient(_Client):
     """
     Overrides all _Client networking functionality to do nothing.
     """

tahoe-lafs 3399.mypy $ tox -e typechecks
typechecks installed: attrs==20.3.0,Automat==20.2.0,cffi==1.14.4,constantly==15.1.0,cryptography==3.3.1,foolscap @ git+https://github.com/jaraco/foolscap@d70d8ecee926d3693b9fac55aebe13573d4f71c9,hyperlink==20.0.1,idna==2.10,incremental==17.5.0,mypy==0.790,mypy-extensions==0.4.3,mypy-zope @ git+https://github.com/jaraco/mypy-zope@48f912cc3cdb4ee37d4d3783ef466a7dce8bf0a9,pyasn1==0.4.8,pyasn1-modules==0.2.8,pycparser==2.20,PyHamcrest==2.0.2,pyOpenSSL==20.0.1,service-identity==18.1.0,six==1.15.0,Twisted==20.3.0,typed-ast==1.4.1,typing-extensions==3.7.4.3,zope.event==4.5.0,zope.interface==5.2.0,zope.schema==6.0.0
typechecks run-test-pre: PYTHONHASHSEED='3588652114'
typechecks run-test: commands[0] | mypy src
src/allmydata/test/no_network.py:264: error: Cannot determine consistent method resolution order (MRO) for "_NoNetworkClient"
Found 1 error in 1 file (checked 293 source files)
ERROR: InvocationError for command /Users/jaraco/code/public/tahoe-lafs/.tox/typechecks/bin/mypy src (exited with code 1)
_______________________________________________________________ summary _______________________________________________________________
ERROR:   typechecks: commands failed
```

I don't understand why the issue exists on some platforms but not others. Probably we need an issue to track this, since it's not consistent.

jaraco replied:
Addressed in 01147f462.

jaraco commented on src/allmydata/test/test_dirnode.py:
```diff
@@ -1587,7 +1587,7 @@ class FakeNodeMaker(NodeMaker):
         return defer.succeed(FakeMutableFile(contents))
 
-class FakeClient2(_Client):
+class FakeClient2(_Client):  # type: ignore # ambiguous MRO
```
Addressed in 01147f462.

jaraco commented on src/allmydata/mutable/filenode.py:
```diff
@@ -564,7 +564,7 @@ def _overwrite(self, new_contents):
 
 
-    def upload(self, new_contents, servermap):
+    def upload(self, new_contents, servermap, progress=None):
```
`blacklist.ProhibitedNode` implements `IFileNode`, which doesn't specify an `upload` method. On the other hand, `mutable.filenode.MutableFileNode` implements `IMutableFileNode`, which does stipulate the interface.

jaraco commented:
In 574613a89, I've merged with master and resolved the conflicts. Doing so revealed 8 new errors:

```
src/allmydata/node.py:302: error: Need type annotation for '_basedir'
src/allmydata/node.py:303: error: Unsupported converter, only named functions and types are currently supported
src/allmydata/test/test_node.py:567: error: "Callable[[TestMissingPorts], Any]" has no attribute "todo"
src/allmydata/test/test_introducer.py:952: error: Dict entry 0 has incompatible type "str": "Dict[int, <nothing>]"; expected "bytes": "Collection[int]"
src/allmydata/test/test_introducer.py:954: error: Dict entry 1 has incompatible type "str": "str"; expected "bytes": "Collection[int]"
src/allmydata/test/web/test_root.py:3: error: Module 'urllib' has no attribute 'quote'
src/allmydata/test/test_storage_client.py:573: error: zope.interface.implementer accepts interface, not foolscap.ipb.IConnectionHintHandler.
src/allmydata/test/test_storage_client.py:573: error: Make sure you have stubs for all packages that provide interfaces for foolscap.ipb.IConnectionHintHandler class hierarchy.
```

jaraco commented:
> [Resolving conflicts] revealed 8 new errors

The five commits through 652222116 address those errors.

jaraco commented:
> * There are some conflicts with master now, alas.

This is addressed (for now).

> * CI isn't completely happy.

With 4a9d3bde5, I believe this is close, if not ready.

> * Are we going to need a ratchet-style job for mypy checks at some point?  It seems like at the end of _this_ PR we have mypy running with no errors but, eg, are a ton of errors going to appear when we bring in the type stubs for our dependencies?  Since that's later it doesn't need to be a blocker for this PR, I'm just wondering - and if there's anything we can do now to make our lives easier then it might be worth it.

I'm not familiar with "ratchet-style job", but my expectation is that there _may_ be some incremental improvements that over time require some maintenance, but overall they should be small. You may be right about there being some maintenance required if foolscap or twisted were to supply stubs, but my instinct is that we can pin on old versions or otherwise degrade the stubs on those projects to maintain status quo if needed.

---

I believe I've largely if not fully addressed all of the concerns above to the extent that I can (notably, merging PRs in foolscap and mypy-zope are out of my control). Please take another look.

jaraco commented:
Now that the blockers are cleared, I recommend to merge this change, and defer any emergent loose ends to follow-up efforts.

exarkun reviewed (approved):
Thanks! Looks good.  Merging.
